### PR TITLE
Add Multi-Tenancy to Circulate

### PR DESCRIPTION
**Pull request by KFad11**
This PR implements the act_as_tenant gem. As well as builds out the groundwork for the multi-tenant structure.

* [x] https://github.com/KFad11/circulate/pull/2
* [x] Acceptance test coverage of logins as users of distinct libraries.
* [x] Data migration to ensure that all existing data gets included in the 0th tenant.
* [x] Scope other necessary models.
* [x] Enforce Chicago postal code only for the Chicago Tool Library - maybe implement `Library#member_postal_code_pattern` to make it generic.
* [x] `Library#logo_image`.

There are some known problems/todo items before this can be effectively used, which will be addressed in a followup PR:

* [x] A Library is created without any members/users, documents, categories, or anything else, and so initial signup and login are not yet possible without manually populating these things like in the seeds.
* [x] An admin of a library can commit a privilege escalation attack, promoting any of their members to super admin. Rules need to be in place preventing role grants beyond the user's own role.
* [ ] A superadmin cannot currently manage the users or any other models belonging to a library other than their own. This requires foregoing the tenant filter on queries when logged in as a superadmin.

None of these items are blockers to merging this PR, which forms a solid baseline for the data model and tenant routing behaviour, but are blockers to effective onboarding of other libraries. That said, we hope to have PRs ready to review to resolve these three independent issues shortly after this one is merged.

### CODE REVIEW

**Comment by benlangfeld** [comment]
@KFad11 @neildecapia https://github.com/KFad11/circulate/pull/1 does some initial scoping. Needs some test coverage and data migration before it's ready to merge.

**Comment by neildecapia** [comment]
@benlangfeld this PR should be ready as our initial push to multi-tenancy. App (in development) is loading and tests are passing.

**Comment by benlangfeld** [comment]
@neildecapia Did you get the system tests working? What did you do?

**Comment by benlangfeld** [comment]
> @benlangfeld this PR should be ready as our initial push to multi-tenancy. App (in development) is loading and tests are passing.

I'd like to get https://github.com/KFad11/circulate/pull/1 completed and merged first. We have some important modifications there, and I'm not keen on scoping Membership like we have done here, without any data migration.

**Comment by neildecapia** [comment]
> @neildecapia Did you get the system tests working? What did you do?

Ran the system tests with `PARALLEL_WORKERS=1` as per #226.

**Comment by benlangfeld** [comment]
> > @neildecapia Did you get the system tests working? What did you do?
> 
> Ran the system tests with `PARALLEL_WORKERS=1` as per #226.

Cool, thanks.

**Comment by benlangfeld** [comment]
@neildecapia The todo items from https://github.com/KFad11/circulate/pull/1 are still pending. I havn't been able to work on circulate today. If @KFad11 gets time, he will look at those, but if not, if would be great if you could pick those up. I've moved them to the description of this PR, along with a few others.

**Comment by neildecapia** [comment]
> @neildecapia The todo items from [KFad11#1](https://github.com/KFad11/circulate/pull/1) are still pending. I havn't been able to work on circulate today. If @KFad11 gets time, he will look at those, but if not, if would be great if you could pick those up. I've moved them to the description of this PR, along with a few others.

@benlangfeld @KFad11 Thanks! I’ll work on these today.

**Comment by benlangfeld** [comment]
@neildecapia took a look through other models and believes that the following need scoping: BorrowPolicy, Category, Document, ShortLink. I can't work on this right now. The test failures are also present on `development`, for what that is worth.

**Comment by neildecapia** [comment]
@benlangfeld @KFad11 Per-library postal code validation: https://github.com/KFad11/circulate/pull/4

**Comment by benlangfeld** [comment]
Thanks @neildecapia . Just needs @KFad11's approval & merge.

**Comment by benlangfeld** [comment]
@neildecapia @KFad11 I can't reproduce that test failure locally. I wonder what might be different in CI.

**Comment by neildecapia** [comment]
@benlangfeld @KFad11 I tried to scope a few more models here: https://github.com/KFad11/circulate/pull/8.

File: db/migrate/20200924072906_establish_tenant_zero.rb
Comment (benlangfeld) [inline db/migrate/20200924072906_establish_tenant_zero.rb:30]: We should add the extra tables from `db/migrate/20200925072325_tenantize_some_more.rb` here and rename this migration to come after that one.

File: db/seeds.rb
Comment (benlangfeld) [inline db/seeds.rb:67]: Do we need to repeat this for the Portland tenant?

**Review by benlangfeld — APPROVED** [review]

**Review by benlangfeld — APPROVED** [review]

**Review by neildecapia — APPROVED** [review]

**Comment by benlangfeld** [comment]
@jim @pollygee Do you have time to review this before we wrap up the week? The two other mentioned deficiencies are actively being addressed.

**Comment by jim** [comment]
I renamed the development branch to `main`, and doing that seems to have closed this PR. I'll open a new one with the changes.

**Pull request merged**
